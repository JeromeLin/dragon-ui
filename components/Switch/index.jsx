import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

class Switch extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.value || props.defaultValue
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: !!nextProps.value
      });
    }
  }

  render() {
    const { props } = this;
    const {
      size, isCheckedText, unCheckedText, isDisabled, style
    } = props;
    const { value } = this.state;
    const disabled = 'disabled' in props || isDisabled;

    const cls = classnames({
      'ui-switch': true,
      checked: value,
      disabled,
      [`size-${size}`]: !!size
    });

    return (
      <span
        className={cls}
        onClick={() => !disabled && this._onClick()}
        style={style}
      >
        <span className="ui-switch-inner">
          {value ? isCheckedText : unCheckedText}
        </span>
      </span>
    );
  }

  _onClick() {
    const value = !this.state.value;
    this.setState({
      value
    });
    this.props.onChange(value);
  }
}

Switch.propTypes = {
  size: PropTypes.oneOf(['sm']),
  value: PropTypes.bool,
  defaultValue: PropTypes.bool,
  isCheckedText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  unCheckedText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  onChange: PropTypes.func
};

Switch.defaultProps = {
  size: null,
  value: false,
  defaultValue: false,
  isCheckedText: '',
  unCheckedText: '',
  onChange: () => {}
};

export default Switch;
