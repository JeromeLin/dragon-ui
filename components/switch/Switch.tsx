import React, { Component, HTMLAttributes } from 'react';
import classnames from 'classnames';
import { PropsType, StateType } from './PropsType';
import Icon from '../icon';

type SwitchPropsType = PropsType & HTMLAttributes<HTMLSpanElement>;

class Switch extends Component<SwitchPropsType, StateType> {
  static defaultProps: PropsType = {
    prefixCls: 'za-switch',
    checked: false,
    defaultChecked: false,
    disabled: false,
    loading: false,
    checkedChildren: '',
    unCheckedChildren: '',
    onChange: () => {},
  };

  state: StateType = {
    checked: this.props.checked || this.props.defaultChecked,
  };

  componentWillReceiveProps(nextProps: SwitchPropsType) {
    if ('checked' in nextProps) {
      this.setState({
        checked: !!nextProps.checked,
      });
    }
  }

  onClick: React.MouseEventHandler<HTMLSpanElement> = (e) => {
    const checked = !this.state.checked;
    this.setState({
      checked,
    });
    this.props.onChange(e, checked);
  }

  render() {
    const { props } = this;
    const {
      size, checkedChildren, unCheckedChildren, prefixCls,
      loading, disabled, className, ...others
    } = props;
    const { checked } = this.state;
    const cls = classnames({
      [prefixCls!]: true,
      [className!]: className,
      [`${prefixCls}--${size}`]: typeof size === 'string' && size,
      [`${prefixCls}--checked`]: checked,
      [`${prefixCls}--disabled`]: disabled || loading,
    });
    const loadingIcon = loading ? <Icon type="loading" className={`${prefixCls}__loadingIcon`} /> : null;
    return (
      <span
        className={cls}
        onClick={(e) => !disabled && !loading && this.onClick(e)}
        {...others}
      >
        {loadingIcon}
        <span className={`${prefixCls}__inner`}>
          {checked ? checkedChildren : unCheckedChildren}
        </span>
      </span>
    );
  }
}

export default Switch;
