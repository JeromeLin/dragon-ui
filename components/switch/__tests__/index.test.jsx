import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Switch from '../Switch';

describe('Switch', () => {
  it('renders normal Switch correctly', () => {
    const wrapper = render(
      <div>
        <Switch />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders Switch correctly with more props', () => {
    const wrapper = render(
      <div>
        <Switch disabled loading size="sm" checkedChildren="是" unCheckedChildren="否" />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('behaves correctly when receving new value', () => {
    const wrapper = mount(
      <Switch defaultChecked={false} />,
    );

    expect(wrapper.find('.za-switch').hasClass('za-switch--checked')).toBeFalsy();
    wrapper.setProps({ checked: true });
    expect(wrapper.find('.za-switch').hasClass('za-switch--checked')).toBeTruthy();
  });

  it('behaves correctly when toggling status', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Switch defaultChecked={false} onChange={onChange} />,
    );

    wrapper.find('.za-switch').simulate('click');
    expect(onChange).toHaveBeenCalledWith(expect.any(Object), true);
  });
});
