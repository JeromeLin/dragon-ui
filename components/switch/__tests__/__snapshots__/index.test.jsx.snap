// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders Switch correctly with more props 1`] = `
<div>
  <span
    class="za-switch za-switch--sm za-switch--disabled"
  >
    <i
      class="za-icon za-icon-loading theme-default za-switch__loadingIcon"
    />
    <span
      class="za-switch__inner"
    >
      Âê¶
    </span>
  </span>
</div>
`;

exports[`Switch renders normal Switch correctly 1`] = `
<div>
  <span
    class="za-switch"
  >
    <span
      class="za-switch__inner"
    />
  </span>
</div>
`;
